apps/emqx_gcp_device/src/emqx_gcp_device.erl

```diff
@@ -0,0 +1,256 @@
+
+-spec create_table() -> ok.
+create_table() ->
+    ok = mria:create_table(?TAB, [
+        {rlog_shard, ?AUTH_SHARD},
+        {type, ordered_set},
+        {storage, disc_copies},
+        {record_name, emqx_gcp_device},
+        {attributes, record_info(fields, emqx_gcp_device)},
+        {storage_properties, [{ets, [{read_concurrency, true}]}]}
+    ]),
+    ok = mria:wait_for_tables([?TAB]).
+
+-spec clear_table() -> ok.
+clear_table() ->
+    {atomic, ok} = mria:clear_table(?TAB),
+    ok.
+
+%%--------------------------------------------------------------------
+%% Internal functions
+%%--------------------------------------------------------------------
+
+-spec transaction(function()) -> term().
+transaction(Fun) ->
+    {atomic, Res} = mria:transaction(?AUTH_SHARD, Fun),
+    Res.
+
+-spec put_device_no_transaction(formatted_device()) -> ok.
+put_device_no_transaction(
+    #{
+        deviceid := DeviceId,
+        keys := Keys,
+        config := EncodedConfig
+    } = Device
+) ->
+    DeviceLoc =
+        list_to_tuple([maps:get(Key, Device, <<>>) || Key <- [project, location, registry]]),
+    ok = put_device(DeviceId, DeviceLoc, Keys),
+    ok = put_config(DeviceId, EncodedConfig).
+
+-spec put_device(deviceid(), device_loc(), [key()]) -> ok.
+put_device(DeviceId, DeviceLoc, Keys) ->
+    CreatedAt = erlang:system_time(second),
+    Extra = #{},
+    Device =
+        #emqx_gcp_device{
+            id = DeviceId,
+            keys = formatted_keys_to_records(Keys),
+            device_loc = DeviceLoc,
+            created_at = CreatedAt,
+            extra = Extra
+        },
+    mria:dirty_write_sync(Device).
+
+-spec formatted_keys_to_records([formatted_key()]) -> [key_record()].
+formatted_keys_to_records(Keys) ->
+    lists:map(fun formatted_key_to_record/1, Keys).
+
+-spec formatted_key_to_record(formatted_key()) -> key_record().
+formatted_key_to_record(#{
+    key_type := KeyType,
+    key := Key,
+    expires_at := ExpiresAt
+}) ->
+    {KeyType, Key, ExpiresAt}.
+
+-spec format_device(emqx_gcp_device()) -> formatted_device().
+format_device(#emqx_gcp_device{
+    id = DeviceId,
+    device_loc = {Project, Location, Registry},
+    keys = Keys,
+    created_at = CreatedAt
+}) ->
+    #{
+        deviceid => DeviceId,
+        project => Project,
+        location => Location,
+        registry => Registry,
+        keys => lists:map(fun format_key/1, Keys),
+        created_at => CreatedAt,
+        config => base64:encode(get_device_config(DeviceId))
+    }.
+
+-spec format_key(key_record()) -> formatted_key().
+format_key({KeyType, Key, ExpiresAt}) ->
+    #{
+        key_type => KeyType,
+        key => Key,
+        expires_at => ExpiresAt
+    }.
+
+-spec put_config(deviceid(), encoded_config()) -> ok.
+put_config(DeviceId, EncodedConfig) ->
+    Config = base64:decode(EncodedConfig),
+    Topic = config_topic(DeviceId),
+    Message = emqx_message:make(undefined, 1, Topic, Config, #{retain => true}, #{}),
```

Comment:
Q: is the `retain` must be true?

Reply:
And why not use the deviceId as the clientid (from) for this message

Reply:
By using the same clientid, [tracing](https://github.com/emqx/emqx/blob/20443267348a753517606e896018687b38339e00/apps/emqx/src/emqx_broker.erl#L217) may pick up those messages.